### ISSUE TITLE
[Bug?]: Hydration error when Element is passed as a prop

### ISSUE THREAD

**Issue by LexSwed** [solidjs/solid-start#1229]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

When passing an element as a prop, Solid fails to hydrate due to key mismatch. Example:
```tsx
<Parent slot={<Child />} />
```

### Expected behavior 🤔

No hydration failure.

### Steps to reproduce 🕹

Steps:

1. Create element that renders `<>{props.slot}</>`
2. Get hydration failure

https://stackblitz.com/edit/solid-start-error-jn34xf?file=src%2Froutes%2Findex.tsx

### Your environment 🌎

```shell
{
  "solid-js": "^1.8.8",
  "vinxi": "^0.1.1",
  "@solidjs/start": "^0.4.4",
}
```

**Comment by thdxr** [comment]
I ran into this as well

**Comment by lxsmnsyc** [comment]
The thing with this one is because you are technically rendering the prop element twice, so it's better that, instead of accessing it a second-time, you utilize Show's callback form.

**Comment by LexSwed** [comment]
Thank you both, after my morning coffee this makes sense now 👍 

Memoizing the access worked (even between server and client 😮), but using callback still produces an error, updated the example:
https://stackblitz.com/edit/solid-start-error-jn34xf?file=src%2Froutes%2Findex.tsx

**Comment by lxsmnsyc** [comment]
@LexSwed Okay interesting, I guess that didn't work. TBF at some point I'd recommend re-imagining the logic for this one. The thing is, anything related to manipulating the JSX could lead to some weird issues (like for the instance of accessing the same props twice).

If that's not something for consideration then I guess the memo solution should work.

**Comment by LexSwed** [comment]
> TBF at some point I'd recommend re-imagining the logic for this one.

Can't imagine giving up on slots. For my simple components it's fine, but for big apps, where the ownership of single UI can be spread between multiple teams, it's best for scaling (e.g. avoiding 300props components -> all the changes they require when something change -> all the conflicts that might happen because of that).

**Comment by danieltroger** [comment]
I think solid-js should warn if elements are created inside of `<Show>`'s `when` - this would be easy to implement in dev mode. Alternatively, `<Show>` could be adapted for this special use case to avoid creating redundant elements.

The memoisation that you did works in most cases but keep in mind that if you, for example, wrap your component before you pass it in `props.slot` in an `<ErrorBoundary>` or another `<Show>`, props.slot would be a getter which always returns a function (contrary to getter that sometimes returns an element, the case you're hitting rn), so your final `<Show>` would go into the "show" branch even if no elements are "in" `props.slot`.

The real solution IMO is using the children helper https://www.solidjs.com/docs/latest#children (which will also get rid of nullish values), so this is guaranteed to work:

```tsx
const resolvedChildren = children(() => props.slot); // Create elements here so they're not created twice + remove nulllish values + flatten resolve nested functions and arrays to a single array of elements
return <Show when={resolvedChildren.toArray().length}><div class="p-2">{resolvedChildren()}</div></Show>
```

**Comment by LexSwed** [comment]
That's very helpful, thanks @danieltroger 🙇  I agree that `children` fits best. Maybe `<Show>` could adapt to that? `when` is a resolved `children`-like, check if it's empty internally?

**Comment by ryansolid** [comment]
> I think solid-js should warn if elements are created inside of <Show>'s when - this would be easy to implement in dev mode.

I don't know if I want to treat `<Show>` as special. In fact I'm pretty sure we don't.

This couldn't easily be done off return type because the children can be dynamic themselves, ie just return functions. We could setup a global that checks if synchronously a `template.cloneNode`, `getNextElement`, or `createComponent` call is made.

I suppose the argument for would be that it would instructional even if someone could fall into this a number of other different ways. In the past I've always restrained from special-casing things because the second you move out of the specific case you haven't solved anything, but I've seen the linter used to great effect with these sort of special case rules. Tricky mind you as linter complaining about multiple prop access for example could yield a lot of false positives.


I'm a bit torn on this in general because I can "fix" this issue in many cases. This started showing up a lot more after 1.8 as far as I figure in that we got stricter. The truth is you never really should use Solid this way, as it is incredibly wasteful, but it used to get past because we would just double render in both envs and match up. I'm tempted to be more permissive and let this through (with a warning) but I'm concerned about the impact of being less strict here. Because we don't really do any rendering during anymore in a lot of cases, like we don't write attributes often things that get by here and do get inserted into the DOM could just cause other weird downstream errors. At least right now it is like this is wrong (cause it is). 

`children` helper is the best tool here. It keeps that knowledge the component author which is where it belongs. If you use a prop twice that is on you, not the person passing props into your component.

**Comment by danieltroger** [comment]
> I don't know if I want to treat `<Show>` as special. In fact I'm pretty sure we don't.

I agree that magically transforming this into a `children` helper call in the scope abow `<Show>` would go to far and not teach the solid-js user about what they're doing. But as for warning, I think the only justification it needs is that many users make a certain mistake, and I've read about people falling into this specific trap and not understanding what's happening multiple times. I think there being a specific dev version of solid js is a great chance here to improve developer experience (warning code wouldn't increase bundle size).

> Tricky mind you as linter complaining about multiple prop access for example could yield a lot of false positives.

Yeah I don't think this would work in the linter, it couldn't be smart enough just based on static analysis of the code, but I might be wrong. Didn't know there was a linter btw, [sounds awesome](https://docs.solidjs.com/guides/how-to-guides/get-ready-for-solid/linting)

> We could setup a global that checks if synchronously a template.cloneNode, getNextElement, or createComponent call is made.

Exactly, I think that's the simplest solution. In solid-js' stack trace tracking style, know if any elements are created in `when` of show and give the users a `children`-using solution to copy paste directly in a warning message. Would that have helped you @LexSwed ?

PS: Sorry I don't quite understand this part of your comment Ryan

> Because we don't really do any rendering during [during what?] anymore in a lot of cases, like we don't write attributes [when?] often things that get by here and do get inserted into the DOM [where in this case do things get inserted into the DOM that shouldn't?] could just cause other weird downstream errors. At least right now it is like this is wrong [what?] (cause it is).

**Comment by ryansolid** [comment]
> Exactly, I think that's the simplest solution. In solid-js' stack trace tracking style, know if any elements are created in when of show and give the users a children-using solution to copy paste directly in a warning message. 

Yeah it's a little bit tricky because of the dance with DOM Expressions currently. Like Solid isn't the only one that needs to be aware we need this mechanism of checking right into the low-level helpers that Solid doesn't provide (that are in the universal library). 

> Do I understand correctly that the current recreation of slot in <Show when={props.slot}>{slot => <div>{slot}</div>}</Show> is expected and the issue can be closed?

You should be aware of this as well https://github.com/solidjs/solid/issues/1977

But I consider this a Solid issue.. not a Start issue.

**Comment by danieltroger** [comment]
> Yeah it's a little bit tricky because of the dance with DOM Expressions currently. Like Solid isn't the only one that needs to be aware we need this mechanism of checking right into the low-level helpers that Solid doesn't provide (that are in the universal library).

Can't you just pick any library to have the state in and then export a `getIsInShowWhen` and `setIsInShowWhen` function?

<details>
<summary>Example</summary>

dom-expressions/dev.js

```tsx
let isInShowWhen = false;
export const getIsInShowWhen = () => isInShowWhen;
export const setIsInShowWhen = (newValue: boolean) => (isInShowWhen = newValue);

// here SSR and element creation functions can interface with isInShowWhen directly
```

solid-js/dev.js:
```tsx
import { setIsInShowWhen } from "dom-expressions";
export function Show<T>(props: { when: T }) {
  const shouldShow = createMemo(() => {
    setIsInShowWhen(true);
    try {
      return props.when;
    } finally {
      setIsInShowWhen(false);
    }
  });
  // rest of logic
}
```
</details>

**Comment by ryansolid** [comment]
Yeah some sort of general mechanism that DOM Expressions exports that is like track creation of elements (as DOM expression has no idea about Show). Honestly this is getting into way too specific range. But I usually collecting these ideas  so that if there is enough reason to add these sort of mechanisms we can. In any case none of this is Start specific. I'm going to move the issue to Solid repo and then close it but that way we can keep the conversation in the right place.
